# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
data = pd.read_csv('data/step3_output.csv')

In [3]:
data.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0      550.000000         1  
1          53.0               90.0      100.395722         1  
2          34.0              152.0       30.000000         1  
3          89.0              122.0      100.395722         0  
4          78.0              104.0       80.000000         0  

[5 rows x 27 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 330 entries, 0 to 329
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               330 non-null    object 
 1   state              330 non-null    object 
 2   summit_elev        330 non-null    int64  
 3   vertical_drop      330 non-null    int64  
 4   base_elev          330 non-null    int64  
 5   trams              330 non-null    int64  
 6   fastEight          330 non-null    float64
 7   fastSixes          330 non-null    int64  
 8   fastQuads          330 non-null    int64  
 9   quad               330 non-null    int64  
 10  triple             330 non-null    int64  
 11  double             330 non-null    int64  
 12  surface            330 non-null    int64  
 13  total_chairs       330 non-null    int64  
 14  Runs               330 non-null    float64
 15  TerrainParks       330 non-null    float64
 16  LongestRun_mi      330 non

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
# making dummis by passing data frame. using this method can dummify several Series
df = pd.get_dummies(data, columns = ['state'], drop_first = True) # dummies are k -1, use drop_first

In [6]:
df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  state_Rhode Island  \
0        0.0          0          2     2       0  ...                   0   
1        0.0          0          0     0       0  ...                   0   
2        0.0          0          0     0       1  ...                   0   
3        0.0          1          0     2       2  ...                   0   
4        0.0          0          1     2       3  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [6]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(x) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [7]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [8]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [9]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [10]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import mean_squared_error
from math import sqrt

In [11]:
rmse = sqrt(mean_squared_error(y_test, y_pred))
rmse

6.897493918032926

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [12]:
lm.intercept_

64.10835351366784

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [13]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(model.coef_), x.columns, columns=['Coefficient'])

Coefficient
summit_elev           2.045083e+01
vertical_drop         6.125083e+00
base_elev             1.843336e+01
trams                 7.316043e+12
fastEight             7.192387e+11
fastSixes             8.514673e+12
fastQuads             2.872207e+13
quad                  1.714529e+13
triple                2.115493e+13
double                2.371446e+13
surface               2.691043e+13
total_chairs          7.576338e+13
Runs                  9.224854e-01
TerrainParks          4.345703e-01
LongestRun_mi         4.794922e-01
SkiableTerrain_ac     1.635742e-01
Snow Making_ac        8.310547e-01
daysOpenLastYear      8.718262e-01
yearsOpen             3.806152e-01
averageSnowfall       3.437500e-01
AdultWeekday          1.923340e+01
projectedDaysOpen     2.294922e-01
NightSkiing_ac        5.058594e-01
clusters              9.335938e-01
state_Arizona         2.158203e-01
state_California      4.155273e-01
state_Colorado        5.273438e-01
state_Connecticut     9.550781e-01
state_Idaho           4.511719e-01
state_Illinois        5.388184e-01
state_Indiana         1.503906e-01
state_Iowa            5.761719e-02
state_Maine           8.754883e-01
state_Maryland        7.456055e-01
state_Massachusetts   2.313477e+00
state_Michigan        7.304688e-01
state_Minnesota       9.765625e-04
state_Missouri        5.224609e-02
state_Montana         9.160156e-01
state_Nevada          2.753906e-01
state_New Hampshire   2.050171e+00
state_New Jersey      3.808594e-01
state_New Mexico      1.036133e+00
state_New York        1.215332e+00
state_North Carolina  2.400391e+00
state_Ohio            3.750000e-01
state_Oregon          2.519531e-01
state_Pennsylvania    1.521484e+00
state_Rhode Island    3.447266e-01
state_South Dakota    1.811523e-01
state_Tennessee       1.185303e+00
state_Utah            5.468750e-01
state_Vermont         1.049316e+00
state_Virginia        1.117188e+00
state_Washington      2.656250e-01
state_West Virginia   1.656738e+00
state_Wisconsin       8.422852e-01
state_Wyoming         4.267578e-01

In [14]:
mean_absolute_error(y_test, y_pred)

5.158801893369886

In [15]:
explained_variance_score(y_test, y_pred)

0.9303237822426702

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [16]:
# Remove state dummies from the df
df1 = df.loc[:, ~df.columns.str.startswith('state')]

In [17]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = df1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df1.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(x) 

In [19]:
# Split into train and test for Model 2

In [18]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [21]:
# Fit Model 2

In [19]:
model = lm.fit(X_train,y_train)
test_model2 = lm.score(X_train,y_train)

In [20]:
y_pred = lm.predict(X_test)

In [21]:
rmse1 = sqrt(mean_squared_error(y_test, y_pred))
rmse1

7.373282783972312

In [22]:
#intercept for model 2
lm.intercept_

64.05683680152094

In [23]:
# Coefficients in DataFrame: 
pd.DataFrame(abs(lm.coef_), x.columns, columns=['Coefficient'])

Coefficient
summit_elev          15.472147
vertical_drop         5.391876
base_elev            12.067335
trams                 0.294501
fastEight             0.053853
fastSixes             0.313608
fastQuads             0.741962
quad                  1.482364
triple                1.307380
double                0.264534
surface               1.299820
total_chairs          0.609203
Runs                  1.595312
TerrainParks          0.169373
LongestRun_mi         0.335457
SkiableTerrain_ac     0.384811
Snow Making_ac        0.480982
daysOpenLastYear      0.681016
yearsOpen             0.200858
averageSnowfall       1.766242
AdultWeekday         19.947600
projectedDaysOpen     0.065078
NightSkiing_ac        0.474623
clusters              0.886656

In [24]:
mean_absolute_error(y_test, y_pred)

5.644428477850815

In [25]:
explained_variance_score(y_test, y_pred)

0.9206185745227597

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [26]:
# Drop Variables
df2 = df1.drop(['summit_elev','base_elev'], axis=1)

In [28]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = df2.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df2.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(x)

In [31]:
# Split into Train and test for model 3

In [29]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [33]:
# Fit Model 3

In [30]:
model = lm.fit(X_train,y_train)

In [31]:
y_pred = lm.predict(X_test)

In [32]:
rmse2 = sqrt(mean_squared_error(y_test, y_pred))
rmse2

7.317769610634036

In [33]:
lm.intercept_

64.05566815635804

In [34]:
# Coefficients in DataFrame: 
pd.DataFrame(abs(lm.coef_), x.columns, columns=['Coefficient'])

Coefficient
vertical_drop         1.374196
trams                 0.263361
fastEight             0.120635
fastSixes             0.289385
fastQuads             0.838049
quad                  1.558961
triple                1.286708
double                0.302736
surface               1.271021
total_chairs          0.681319
Runs                  1.772081
TerrainParks          0.290070
LongestRun_mi         0.333990
SkiableTerrain_ac     0.451886
Snow Making_ac        0.429863
daysOpenLastYear      0.651870
yearsOpen             0.192631
averageSnowfall       2.211218
AdultWeekday         19.899294
projectedDaysOpen     0.238347
NightSkiing_ac        0.484379
clusters              0.692033

In [36]:
mean_absolute_error(y_test, y_pred)

5.696125384548971

In [37]:
explained_variance_score(y_test, y_pred)

0.9218195823124583

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. |0.9303237822426702|5.158801893369886|-|
| Model 2. | 0.9206185745227597|5.644428477850815 |'state'|
| Model 3. | 0.9218195823124583 |5.696125384548971|'state','summit_elev','base_elev'|

Model Selection:

In [41]:
# Based on the Explained Variance, and Mean Absolute Error Model 1 is performing well than the two other.
# Hence, Model 1 is selected